gordon/router.py

```diff
@@ -0,0 +1,162 @@
+provider if no enricher provider is installed).
+
+If a message fails its next phase, its phase will be updated to ``drop``
+and routed to the event consumer provider for cleanup.
+
+.. attention::
+
+    The :class:`GordonRouter` only supports the following two phase
+    routes:
+
+    1. consume -> enrich -> publish -> done
+    2. consume -> publish -> done
+
+    Future releases may support more configurable phase routings.
+
+"""
+
+import asyncio
+import logging
+
+
+# TODO/NOTE (lynn): This is a temporary helper function to alias
+#                   entrypoints of plugins to a common name. In the
+#                   future, plugin interfaces will be updated to have a
+#                   common name for their entrypoint.
+def _set_plugin_entrypoint_aliases(plugin_dict):
+    # e.g. enricher_inst.process => enricher_inst.handle_message
+    entrypoints = {
+        'enricher': 'process',
+        'publisher': 'publish_changes',
+        'event_consumer': 'cleanup',
+    }
+    for plugin_name, func in entrypoints.items():
+        try:
+            plugin = plugin_dict[plugin_name]
+        except KeyError:  # if no enricher
+            continue
+        entry_point = getattr(plugin, func)
+        setattr(plugin, 'handle_message', entry_point)
+    return plugin_dict
+
+
+class GordonRouter:
+    """Route messages to the appropriate plugin destination.
+
+    .. attention::
+
+        `error_channel` is currently not used in this router, and may
+        be removed entirely from all interface definitions.
+
+    Args:
+        success_channel (asyncio.Queue): A sink for successfully
+            processed :class:`gordon.interfaces.IEventMessage` s.
+        error_channel (asyncio.Queue): A sink for
+            :class:`gordon.interfaces.IEventMessage` s that were not
+            processed due to problems.
+        plugins (dict): Implemented plugin provider names mapped to
+            their instantiated objects.
+    """
+    # TODO (lynn): Ideally this is configurable/extendable in future
+    #              iterations of gordon.
+    PHASE_TO_PLUGIN_MAPPER = {
+        'enrich': 'enricher',
+        'publish': 'publisher',
+        'cleanup': 'event_consumer',
+    }
+    FINAL_PHASES = ('cleanup',)
+
+    def __init__(self, success_channel, error_channel, plugins):
+        self.success_channel = success_channel
+        self.error_channel = error_channel
+        self.plugins = _set_plugin_entrypoint_aliases(plugins)
+        self.phase_route = self._setup_phase_route()
+
+    # TODO (lynn): Ideally this is configurable/extendable in future
+    #              iterations of gordon.
+    def _setup_phase_route(self):
+        path = {
+            'consume': 'publish',
+            'publish': 'cleanup',
+            'cleanup': 'cleanup',
+        }
+        if 'enricher' in self.plugins:
+            path['consume'] = 'enrich'
+            path['enrich'] = 'publish'
+        return path
```

Comment:
Tell me if I'm being too picky since it's just an local var, but could we call it `route` as well?